`pytest` on current dev version

**Describe the issue**
When running `pytest` with on the current development version of `sisl` on my machine I get lots of failures, for example:
```
$ SISL_FILES_TESTS=$(pwd)/sisl-files/tests pytest-3 --pyargs sisl._core.tests.test_sparse
============================= test session starts ============================
==
platform linux -- Python 3.10.12, pytest-8.0.2, pluggy-1.4.0
rootdir: /home/me
plugins: flake8-1.0.6, pylint-0.18.0
collected 88 items

. ................F....FFFFFFFF.................x.............FFFFFFFFFF [ 79%
]
FFF.FF...........F                                                       [100%
]

=================================== FAILURES =================================
==
______________________________ test_diags_offsets ____________________________
__

    def test_diags_offsets():
        csr = SparseCSR((10, 10), dtype=np.int32)
        m = csr.diags(1)
        assert m.nnz == 10
>       assert np.sum(m) == 10

../../.local/lib/python3.10/site-packages/sisl/_core/tests/test_sparse.py:225:
 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
_ 
<__array_function__ internals>:5: in sum
    ???
/usr/lib/python3/dist-packages/numpy/core/fromnumeric.py:2259: in sum
    return _wrapreduction(a, np.add, 'sum', axis, dtype, out, keepdims=keepdim
s,
/usr/lib/python3/dist-packages/numpy/core/fromnumeric.py:86: in _wrapreduction
    return ufunc.reduce(obj, axis, dtype, out, **passkwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
_ 

self = <sisl.SparseCSR shape=(10, 10, 1), kind=i, nnz=10>, ufunc = <ufunc 'add
'>
method = 'reduce'
inputs = (<sisl.SparseCSR shape=(10, 10, 1), kind=i, nnz=10>,)
kwargs = {'axis': None, 'dtype': None}, out = None
dtypes = [None, dtype('int64'), None]
arg = <sisl.SparseCSR shape=(10, 10, 1), kind=i, nnz=10>

    def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
        out = kwargs.pop("out", None)
    
        if getattr(ufunc, "signature", None) is not None:
            # The signature is not a scalar operation
            return NotImplemented
    
        dtypes = []
        if out is not None:
            (out,) = out
            kwargs["dtype"] = out.dtype
        else:
            for arg in inputs:
                # this exercise is required to ensure that we don't
                # prematurely promote dtypes.
                # For instance,
                #    SparseCSR(dtype=np.complex64) * 1j
                # should remain np.complex64.
                # Yet, if one does np.asarray(1j) you'll get
                # np.complex128.
                # This means that we should let the ufunc do any
                # promotions necessary.
                if isscalar(arg):
                    pass
                elif isinstance(arg, (tuple, list, ndarray)):
                    arg = arg[0]
                try:
                    dtypes.append(arg.dtype)
                except AttributeError:
                    dtypes.append(type(arg))
    
        if method == "__call__":
            if dtypes:
                dtypes.append(None)
                kwargs["dtype"] = ufunc.resolve_dtypes(tuple(dtypes))[-1]
            result = _ufunc_call(ufunc, *inputs, **kwargs)
        elif method == "reduce":
            if dtypes:
                dtypes = [None] + dtypes + [None]
>               kwargs["dtype"] = ufunc.resolve_dtypes(tuple(dtypes), reductio
n=True)[
                    -1
                ]
E               AttributeError: 'numpy.ufunc' object has no attribute 'resolve
_dtypes'

../../.local/lib/python3.10/site-packages/sisl/_core/sparse.py:1894: Attribute
Error
...
```
I suppose some of package dependencies in my environment need to be updated. However, I raise this issue here because I got no errors or warnings during the sisl installation (requirements satisfied?), and I have not encountered any apparent issues running different sisl scripts that I use.
```
$ python3 -m pip install . --user
Processing /home/me/sisl
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... done
Requirement already satisfied: numpy>=1.13 in /usr/lib/python3/dist-packages (from sisl==0.14.4.dev182+g2334b58e6) (1.21.5)
Requirement already satisfied: scipy>=1.5 in /usr/lib/python3/dist-packages (from sisl==0.14.4.dev182+g2334b58e6) (1.8.0)
Requirement already satisfied: pyparsing>=1.5.7 in /usr/lib/python3/dist-packages (from sisl==0.14.4.dev182+g2334b58e6) (2.4.7)
Requirement already satisfied: xarray>=0.10 in /usr/lib/python3/dist-packages (from sisl==0.14.4.dev182+g2334b58e6) (0.16.1)
Building wheels for collected packages: sisl
  Building wheel for sisl (pyproject.toml) ... done
  Created wheel for sisl: filename=sisl-0.14.4.dev182+g2334b58e6-cp310-cp310-manylinux_2_35_x86_64.whl size=4939789 sha256=789685a8243e31fd09062ebfb421781c8d45a5cb7a64bc6906778d2e9a10429d
  Stored in directory: /tmp/pip-ephem-wheel-cache-iwjk04sg/wheels/f8/d5/99/60fdf65fb542d140b08984c193ce9f7cbc37f326ab3a52d94c
Successfully built sisl
Installing collected packages: sisl
  Attempting uninstall: sisl
    Found existing installation: sisl 0.14.4.dev182+g2334b58e6
    Uninstalling sisl-0.14.4.dev182+g2334b58e6:
      Successfully uninstalled sisl-0.14.4.dev182+g2334b58e6
Successfully installed sisl-0.14.4.dev182+g2334b58e6
```


**Version details**
Run the below code and add to issue (if an issue is relevant for the issue):
```python
>>> sys.version
'3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0]'
>>> sisl.__version__
'0.14.4.dev182+g2334b58e6'
```

With `700-numpy` I get different errors:
```
============================= test session starts ==============================
platform linux -- Python 3.10.12, pytest-8.0.2, pluggy-1.4.0
rootdir: /home/me
plugins: flake8-1.0.6, pylint-0.18.0
collected 88 items

. ..............................................x...............FFF..... [ 79%]
.F................                                                       [100%]

=================================== FAILURES ===================================
___________________________________ test_op3 ___________________________________

    def test_op3():
        S = SparseCSR((10, 100), dtype=np.int32)
        # Create initial stuff
        for i in range(10):
            j = range(i * 4, i * 4 + 3)
            S[0, j] = i
    
        for op in ["add", "sub", "mul", "pow"]:
            func = getattr(S, f"__{op}__")
            s = func(1)
            # 1 will promote to whatever numpy will cast it to
>           assert s.dtype == np.int32
E           AssertionError: assert dtype('float64') == <class 'numpy.int32'>
E            +  where dtype('float64') = <sisl.SparseCSR shape=(10, 100, 1), kind=f, nnz=30>.dtype
E            +  and   <class 'numpy.int32'> = np.int32

../../.local/lib/python3.10/site-packages/sisl/_core/tests/test_sparse.py:1043: AssertionError
___________________________________ test_op4 ___________________________________

    def test_op4():
        S = SparseCSR((10, 100), dtype=np.int32)
        # Create initial stuff
        for i in range(10):
            j = range(i * 4, i * 4 + 3)
            S[0, j] = i
    
        s = 1 + S
>       assert s.dtype == np.int32
E       AssertionError: assert dtype('float64') == <class 'numpy.int32'>
E        +  where dtype('float64') = <sisl.SparseCSR shape=(10, 100, 1), kind=f, nnz=30>.dtype
E        +  and   <class 'numpy.int32'> = np.int32

../../.local/lib/python3.10/site-packages/sisl/_core/tests/test_sparse.py:1081: AssertionError
_____________________________ test_op_binary[mod] ______________________________

matrix_sisl_csr = [<sisl.SparseCSR shape=(10, 80, 1), kind=i, nnz=10>, <sisl.SparseCSR shape=(10, 80, 1),
 kind=i, nnz=10>, <sisl.SparseC...d=i, nnz=7>, <sisl.SparseCSR shape=(10, 80, 1), kind=i, nnz=7>, <sisl.S
parseCSR shape=(10, 80, 1), kind=i, nnz=7>, ...]
matrix_csr_matrix = [<10x80 sparse matrix of type '<class 'numpy.int32'>'
	with 10 stored elements in Compressed Sparse Row format>, <10x8...ormat>, <10x80 sparse matrix of
 type '<class 'numpy.int32'>'
	with 37 stored elements in Compressed Sparse Row format>]
--More--
```

> Hmm. I would like to see how your environment looks like with numpy.
> 
> I.e. I tried installing numpy==1.24.0, and everything works. Do you use virtual environments or something like that?

No, no virtual environment.

> You also run `pytest-3`, could you do:
> 
> ```shell
> which python3
> which pytest
> which sgeom
> python3 -c "import numpy as np ; print(np.__version__)"
> ```

```
which python3
/usr/bin/python3
$ which pytest
/home/me/.local/bin/pytest
$ which pytest-3
/usr/bin/pytest-3
$ which sgeom
/home/me/.local/bin/sgeom
$ python3 -c "import numpy as np ; print(np.__version__)"
1.24.0
```
I notice here that apparently I have both `pytest` and `pytest-3` and in different locations. 

Comparing the outputs from these I see some differences (but failures with both):
```
$ SISL_FILES_TESTS=$(pwd)/sisl-files/tests pytest-3 --pyargs sisl._core > test-3
$ SISL_FILES_TESTS=$(pwd)/sisl-files/tests pytest --pyargs sisl._core > test
$ diff test-3 test | tail
< 
< ../../.local/lib/python3.10/site-packages/sisl/_core/tests/test_sparse.py:1297: AssertionError
164,168c81
< FAILED tests/test_sparse.py::test_op3 - AssertionError: assert dtype('float64...
< FAILED tests/test_sparse.py::test_op4 - AssertionError: assert dtype('float64...
< FAILED tests/test_sparse.py::test_op_binary[mod] - assert False
< FAILED tests/test_sparse.py::test_op_numpy_scalar - AssertionError: assert dt...
< == 5 failed, 659 passed, 1 skipped, 1 xfailed, 9 warnings in 78.09s (0:01:18) ==
---
> == 1 failed, 663 passed, 1 skipped, 1 xfailed, 9 warnings in 78.45s (0:01:18) ==
```

The only one picked up with `pytest` is this one:
```
=================================== FAILURES ===================================
_________________________ TestGeometry.test_distance8 __________________________

self = <sisl._core.tests.test_geometry.TestGeometry object at 0x7f82c8498d90>
setup = <sisl._core.tests.test_geometry.setup.<locals>.t object at 0x7f82c60eabc0>

    def test_distance8(self, setup):
        geom = Geometry([0] * 3, Atom(1, R=1.0), lattice=1)
        geom.set_nsc([77, 1, 1])
        d = geom.distance(0, method="min")
        assert len(d) == 1
        d = geom.distance(0, method="median")
        assert len(d) == 1
>       d = geom.distance(0, method="mode")

../../.local/lib/python3.10/site-packages/sisl/_core/tests/test_geometry.py:1254: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../.local/lib/python3.10/site-packages/sisl/_core/geometry.py:3406: in distance
    shells[i] = func(lst)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

lst = [1.0, 1.0]

    def func(lst):
>       return mode(lst, keepdims=False)[0]
E       TypeError: mode() got an unexpected keyword argument 'keepdims'

../../.local/lib/python3.10/site-packages/sisl/_core/geometry.py:3387: TypeError
=============================== warnings summary ===============================
../../../../usr/lib/python3/dist-packages/blosc/toplevel.py:12
  /usr/lib/python3/dist-packages/blosc/toplevel.py:12: DeprecationWarning: The distutils package is deprecated and slated for removal in Python 3.12. Use setuptools or check PEP 632 for potential alternatives
    from distutils.version import LooseVersion

tests/test_geometry.py::TestGeometry::test_close1
tests/test_geometry.py::TestGeometry::test_close2
tests/test_geometry.py::TestGeometry::test_close_within1
tests/test_geometry.py::TestGeometry::test_bond_correct
  warn:0: SislWarning: Geometry.close_sc has been passed an 'atoms' argument together with an R value larger than the orbital ranges. If used together with 'sparse-matrix.construct' this can result in wrong coup
lings.

tests/test_grid.py::TestGrid::test_pyamg2
  warn:0: SislWarning: Lattice.set_boundary_condition is having image connections (nsc=3>1) while having a non-periodic boundary condition.

tests/test_sparse_geometry.py::TestSparseAtom::test_untile_wrong_usage
  warn:0: SislWarning: SparseAtom.untile may have connections crossing the entire unit cell. This may result in wrong behavior due to non-unique matrix elements.

tests/test_sparse_geometry.py::test_sparse_orbital_add_axis
tests/test_sparse_geometry.py::test_sparse_orbital_add_no_axis
  warn:0: SislWarning: Lattice got initialized with one or more lattice vector(s) with 0 length. Use with care.

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== short test summary info ============================
FAILED tests/test_geometry.py::TestGeometry::test_distance8 - TypeError: mode...
== 1 failed, 663 passed, 1 skipped, 1 xfailed, 9 warnings in 78.45s (0:01:18) ==

```

Stepping back to 1.23.5 brings back one additional failure (4 in total):
```
geometry = <sisl.Geometry na=2, no=2, nsc=[3 3 3]>

    def test_add_xyz_to_bonds_dataset(geometry):
        bonds = find_all_bonds(geometry, 1.5)
    
        xyz_bonds = add_xyz_to_bonds_dataset(bonds)
    
        assert isinstance(xyz_bonds, xr.Dataset)
        assert "xyz" in xyz_bonds.data_vars
        assert xyz_bonds.xyz.shape == (len(bonds.bonds), 2, 3)
>       assert np.allclose(xyz_bonds.xyz[:, 0], geometry.xyz[bonds.bonds[:, 0]])
E       IndexError: too many indices for array: array is 2-dimensional, but 27 were indexed

../../.local/lib/python3.10/site-packages/sisl/viz/processors/tests/test_geometry.py:272: IndexError
```

--- a/ci/requirements.txt
+++ b/ci/requirements.txt
@@ -1,5 +1,4 @@
-# We use some functionality in 1.13 (np.unique)
-numpy>=1.13
+numpy>=1.19.5
 scipy>=1.5
 pyparsing>=1.5.7
 xarray>=0.10

--- a/docs/installation.rst
+++ b/docs/installation.rst
@@ -12,7 +12,7 @@ Required dependencies
 Running sisl requires these versions:
 
 - `Python`_ 3.8 or above
-- `numpy`_ (1.13 or later)
+- `numpy`_ (1.19.5 or later)
 - `scipy`_ (1.5 or later)
 - `pyparsing`_ (1.5.7 or later)
 - `xarray`_ (0.10.0 or later)
@@ -213,4 +213,3 @@ A basic procedure would be:
 
    git clone https://github.com/zerothi/sisl-files.git
    SISL_FILES_TESTS=$(pwd)/sisl-files/tests pytest --pyargs sisl
-

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -42,7 +42,7 @@ dependencies = [
     # TODO work-arounds for windows installations
     "numpy>=1.19.5 ; sys_platform == 'win32' and python_version <= '3.9'",
     "numpy>=1.21.5 ; sys_platform == 'win32' and python_version >= '3.10'",
-    "numpy>=1.13 ; sys_platform != 'win32'",
+    "numpy>=1.19.5 ; sys_platform != 'win32'",
     "scipy>=1.5",
     "pyparsing>=1.5.7",
     "xarray>=0.10",

--- a/src/sisl/_core/sparse.py
+++ b/src/sisl/_core/sparse.py
@@ -1859,41 +1859,12 @@ column indices of the sparse elements
             # The signature is not a scalar operation
             return NotImplemented
 
-        dtypes = []
         if out is not None:
             (out,) = out
-            kwargs["dtype"] = out.dtype
-        else:
-            for arg in inputs:
-                # this exercise is required to ensure that we don't
-                # prematurely promote dtypes.
-                # For instance,
-                #    SparseCSR(dtype=np.complex64) * 1j
-                # should remain np.complex64.
-                # Yet, if one does np.asarray(1j) you'll get
-                # np.complex128.
-                # This means that we should let the ufunc do any
-                # promotions necessary.
-                if isscalar(arg):
-                    pass
-                elif isinstance(arg, (tuple, list, ndarray)):
-                    arg = arg[0]
-                try:
-                    dtypes.append(arg.dtype)
-                except AttributeError:
-                    dtypes.append(type(arg))
 
         if method == "__call__":
-            if dtypes:
-                dtypes.append(None)
-                kwargs["dtype"] = ufunc.resolve_dtypes(tuple(dtypes))[-1]
             result = _ufunc_call(ufunc, *inputs, **kwargs)
         elif method == "reduce":
-            if dtypes:
-                dtypes = [None] + dtypes + [None]
-                kwargs["dtype"] = ufunc.resolve_dtypes(tuple(dtypes), reduction=True)[
-                    -1
-                ]
             result = _ufunc_reduce(ufunc, *inputs, **kwargs)
         elif method == "outer":
             # Currently I don't know what to do here
@@ -1974,6 +1945,14 @@ def _ufunc(ufunc, a, b, **kwargs):
     return ufunc(a, b, **kwargs)
 
 
+def _sp_data_cast(sp, result, dtype=None):
+    """Converts the data in `sp` to the result data-type"""
+    if dtype is None:
+        dtype = result.dtype
+    if sp.dtype != dtype:
+        sp._D = sp._D.astype(dtype)
+
+
 def _ufunc_sp_ndarray(ufunc, a, b, **kwargs):
     if len(_get_reduced_shape(b)) > 1:
         # there are shapes for individiual
@@ -1981,7 +1960,7 @@ def _ufunc_sp_ndarray(ufunc, a, b, **kwargs):
         return ufunc(a.toarray(), b, **kwargs)
 
     # create a copy
-    out = a.copy(dtype=kwargs["dtype"])
+    out = a.copy(dtype=kwargs.get("dtype", None))
     if out.ptr[-1] == out.nnz:
         out._D = ufunc(a._D, b, **kwargs)
     else:
@@ -1990,7 +1969,9 @@ def _ufunc_sp_ndarray(ufunc, a, b, **kwargs):
         # a view, we can't use
         #   ufunc(..., out=out._D[idx, :])
         idx = array_arange(a.ptr[:-1], n=a.ncol)
-        out._D[idx] = ufunc(a._D[idx, :], b, **kwargs)
+        D = ufunc(a._D[idx, :], b, **kwargs)
+        _sp_data_cast(out, D)
+        out._D[idx] = D
         del idx
     return out
 
@@ -2002,7 +1983,7 @@ def _ufunc_ndarray_sp(ufunc, a, b, **kwargs):
         return ufunc(a, b.toarray(), **kwargs)
 
     # create a copy
-    out = b.copy(dtype=kwargs["dtype"])
+    out = b.copy(dtype=kwargs.get("dtype", None))
     if out.ptr[-1] == out.nnz:
         out._D = ufunc(a, b._D, **kwargs)
     else:
@@ -2011,7 +1992,9 @@ def _ufunc_ndarray_sp(ufunc, a, b, **kwargs):
         # a view, we can't use
         #   ufunc(..., out=out._D[idx, :])
         idx = array_arange(b.ptr[:-1], n=b.ncol)
-        out._D[idx] = ufunc(a, b._D[idx, :], **kwargs)
+        D = ufunc(a, b._D[idx, :], **kwargs)
+        _sp_data_cast(out, D)
+        out._D[idx] = D
         del idx
     return out
 
@@ -2065,7 +2048,12 @@ def _ufunc_sp_sp(ufunc, a, b, **kwargs):
         raise ValueError(f"could not broadcast sparse matrices {a.shape} and {b.shape}")
 
     # create union of the sparsity pattern
-    out = SparseCSR.sparsity_union(a, b, dim=max(adim, bdim), dtype=kwargs["dtype"])
+    # create a fake *out* to grap the dtype
+    dtype = kwargs.get("dtype")
+    if dtype is None:
+        out = ufunc(adata[:1, :], bdata[:1, :], **kwargs)
+        dtype = out.dtype
+    out = SparseCSR.sparsity_union(a, b, dim=max(adim, bdim), dtype=dtype)
 
     for r in range(out.shape[0]):
         offset = out.ptr[r]
@@ -2118,13 +2106,15 @@ def _ufunc_call(ufunc, *in_args, **kwargs):
     if len(args) == 1:
         a = args[0]
         # create a copy
-        out = a.copy(dtype=kwargs["dtype"])
+        out = a.copy(dtype=kwargs.get("dtype", None))
         if out.ptr[-1] == out.nnz:
             out._D = ufunc(a._D, **kwargs)
         else:
             # limit the values
             idx = array_arange(a.ptr[:-1], n=a.ncol)
-            out._D[idx] = ufunc(a._D[idx, :], **kwargs)
+            D = ufunc(a._D[idx, :], **kwargs)
+            _sp_data_cast(out, D)
+            out._D[idx] = D
             del idx
         return out
 
@@ -2171,15 +2161,18 @@ def _ufunc_reduce(ufunc, array, axis=0, *args, **kwargs):
     elif axis == 1:
         pass
     elif axis == 2:
-        out = array.copy(dims=range(array.shape[2] - 1), dtype=kwargs["dtype"])
-        out._D[:, 0] = ufunc.reduce(array._D, axis=1, *args, **kwargs)
+        out = array.copy(dims=0, dtype=kwargs.get("dtype", None))
+        D = ufunc.reduce(array._D, axis=1, *args, **kwargs)
+        _sp_data_cast(out, D, kwargs.get("dtype"))
+        out._D[:, 0] = D
         return out
     else:
         raise ValueError(
             f"Unknown axis argument in ufunc.reduce call on {array.__class__.__name__}"
         )
 
-    ret = empty([array.shape[0], array.shape[2]], dtype=kwargs["dtype"])
+    ret = ufunc.reduce(array._D[0:1, :], axis=0, *args, **kwargs)
+    ret = empty([array.shape[0], array.shape[2]], dtype=kwargs.get("dtype", ret.dtype))
 
     # Now do ufunc calculations, note that initial gets passed directly
     ptr = array.ptr

--- a/src/sisl/_core/tests/test_sparse.py
+++ b/src/sisl/_core/tests/test_sparse.py
@@ -1036,10 +1036,9 @@ def test_op3():
         j = range(i * 4, i * 4 + 3)
         S[0, j] = i
 
-    for op in ["add", "sub", "mul", "pow"]:
+    for op in ("add", "sub", "mul", "pow"):
         func = getattr(S, f"__{op}__")
         s = func(1)
-        # 1 will promote to whatever numpy will cast it to
         assert s.dtype == np.int32
         s = func(1.0)
         assert s.dtype == np.float64
@@ -1048,7 +1047,7 @@ def test_op3():
             assert s.dtype == np.complex128
 
     S = S.copy(dtype=np.float64)
-    for op in ["add", "sub", "mul", "pow"]:
+    for op in ("add", "sub", "mul", "pow"):
         func = getattr(S, f"__{op}__")
         s = func(1)
         assert s.dtype == np.float64
@@ -1059,7 +1058,7 @@ def test_op3():
             assert s.dtype == np.complex128
 
     S = S.copy(dtype=np.complex128)
-    for op in ["add", "sub", "mul", "pow"]:
+    for op in ("add", "sub", "mul", "pow"):
         func = getattr(S, f"__{op}__")
         s = func(1)
         assert s.dtype == np.complex128

--- a/src/sisl/physics/densitymatrix.py
+++ b/src/sisl/physics/densitymatrix.py
@@ -692,17 +692,6 @@ class _densitymatrix(SparseOrbitalBZSpin):
         eta : bool, optional
            show a progressbar on stdout
         """
-        try:
-            # Once unique has the axis keyword, we know we can safely
-            # use it in this routine
-            # Otherwise we raise an ImportError
-            unique([[0, 1], [2, 3]], axis=0)
-        except Exception:
-            raise NotImplementedError(
-                f"{self.__class__.__name__}.density requires numpy >= 1.13, either update "
-                "numpy or do not use this function!"
-            )
-
         geometry = self.geometry
         # Check that the atomic coordinates, really are all within the intrinsic supercell.
         # If not, it may mean that the DM does not conform to the primary unit-cell paradigm
@@ -953,7 +942,6 @@ class _densitymatrix(SparseOrbitalBZSpin):
             idx[idx[:, 2] < 0, 2] = 0
             idx[shape[2] <= idx[:, 2], 2] = shape[2] - 1
 
-            # Remove duplicates, requires numpy >= 1.13
             idx = unique(idx, axis=0)
             if len(idx) == 0:
                 eta.update()

--- a/src/sisl/physics/state.py
+++ b/src/sisl/physics/state.py
@@ -1119,8 +1119,6 @@ coefficients assigned to each state
                 if matrix:
                     ret = (v,)
                 else:
-                    # numpy >= 1.9 returns a read-only view of the data,
-                    # so take a copy to ensure editable state
                     ret = (np.diagonal(v, axis1=1, axis2=2).copy(),)
 
             else:

